### PR TITLE
Coinbase schedule and paying-forward

### PR DESCRIPTION
**Pull request by kantai**
## Description

This PR implements the proposed coinbase schedule in #1888 and the pay-forward of missed sortitions scheme proposed in #2004. This includes the initial block mining bonus behavior, where any burnchain blocks between the first burn block height and the first sortition winner pay their coinbases out to an initial block window (constant set in `core/mod.rs`).

## Type of Change
- [x] New feature

## Does this introduce a breaking change?
Yes.

## Are documentation updates required?
No -- documentation on coinbase rewards already reflect this behavior.

## Testing information

An integration test is included in the coordinator::tests module that tests both the initial block mining bonus and the paying forward of missing sortitions.

Many of the existing integration tests needed to be updated with the new coinbase reward schedule, but especially the initial mining bonus.

### CODE REVIEW

File: src/chainstate/stacks/db/blocks.rs [hunk 1]
```diff
@@ -4170,6 +4183,22 @@ impl StacksChainState {
...
+            )?
+            .expect("CORRUPTION: failed to load snapshot that elected processed block")
+            .accumulated_coinbase_ustx;
+
+            let coinbase_at_block = StacksChainState::get_coinbase_reward(
+                chain_tip_burn_header_height as u64,
+                burn_dbconn.context.first_block_height,
+            );
+
+            let total_coinbase = coinbase_at_block + accumulated_rewards;
```
Comment (jcnelson) [inline src/chainstate/stacks/db/blocks.rs:4200]: Maybe do a saturating add here?

File: src/chainstate/burn/db/sortdb.rs [hunk 2]
```diff
@@ -3382,6 +3438,31 @@ impl<'a> SortitionHandleTx<'a> {
             values.push(snapshot.sortition_id.to_hex());
         }
 
+        if let Some(initialize_bonus) = initialize_bonus {
+            // first sortition with a winner, set the initial mining bonus fields
+            keys.push(db_keys::initial_mining_bonus_per_block().into());
+            values.push(initialize_bonus.per_block.to_string());
+
+            let total_reward_remaining = initialize_bonus.total_reward - initialize_bonus.per_block;
```
Comment (jcnelson) [inline src/chainstate/burn/db/sortdb.rs:3446]: Is there a case where this could be negative?  Should we use a saturating sub to be safe?

**Review by jcnelson — APPROVED** [review]
This LGTM.  Thanks @kantai!  Just please address the saturating add/sub comments before merging.

File: src/core/mod.rs [hunk 3]
```diff
@@ -39,6 +39,14 @@ pub const NETWORK_P2P_PORT: u16 = 6265;
 
 pub const MINING_COMMITMENT_WINDOW: u8 = 6;
 
+/// The number of blocks which will share the block bonus
+///   from burn blocks that occurred without a sortition.
+///   (See: https://forum.stacks.org/t/pox-consensus-and-stx-future-supply)
+#[cfg(test)]
+pub const INITIAL_MINING_BONUS_WINDOW: u16 = 10;
+#[cfg(not(test))]
+pub const INITIAL_MINING_BONUS_WINDOW: u16 = 10_000;
```
Comment (diwakergupta) [inline src/core/mod.rs:48]: Does it make sense to make this a multiple of the PoX reward cycle length? That way we can ensure that the mining bonus couldn't potentially interfere with Stacking mid-cycle

Comment (kantai) [reply src/core/mod.rs:48]: Yeah, I think that's probably a good idea -- a related change would need to be made to the way we calculate the PoX reward cycle. Right now, we simply take `(burn_block_height - first_burn_block) % reward_cycle_length`. But the initial mining bonus window starts at `first_burn_block_with_active_mining` (the difference between `first_burn_block` and `first_burn_block_with_active_mining` is what builds up that initial bonus).

File: src/chainstate/burn/db/sortdb.rs (hunk 2 above)
Comment (kantai) [reply src/chainstate/burn/db/sortdb.rs]: This couldn't be negative, as `per_block` is set as just `total_reward / INITIAL_MINING_WINDOW`, but I updated to saturating sub just to be on the safe side in ce026c8a1.

File: src/chainstate/stacks/db/blocks.rs (hunk 1 above)
Comment (kantai) [reply src/chainstate/stacks/db/blocks.rs]: Updated in ce026c8a1

**Review by lgalabru — APPROVED** [review]
Looks good to me, thank you @kantai !

File: src/core/mod.rs (hunk 3 above)
Comment (kantai) [reply src/core/mod.rs:48]: Opened a new issue for tracking this discussion: #2140.

**Pull request merged**